File: packages/dcf_reanimated/ios/Classes/Components/DCFCanvasComponent.swift
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import UIKit
import Flutter
import dcflight

/// Pure Canvas view - displays Flutter/Skia textures
/// All rendering (particles, shaders, paths, etc.) happens in Skia-land
/// Native side is just a texture container - no hardcoded primitives
class DCFCanvasView: UIView {
    // Flutter texture display - renders whatever Skia draws
    private var imageView: UIImageView?
    
    // Metadata
    var nodeId: String?
    var canvasId: String?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .clear
        isOpaque = false
        
        // Create image view for Flutter/Skia texture display
        imageView = UIImageView()
        imageView?.contentMode = .scaleToFill
        if let imageView = imageView {
            addSubview(imageView)
        }
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        imageView?.frame = bounds
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    /// Update Flutter/Skia texture
    /// Everything is rendered in Skia-land: particles, shaders, paths, animations, etc.
    /// Native just displays the texture - no hardcoded primitives
    func updateTexture(with imageData: Data, width: Int, height: Int) {
        guard let image = createImage(from: imageData, width: width, height: height) else {
            return
        }
        
        DispatchQueue.main.async { [weak self] in
            self?.imageView?.image = image
        }
    }
    
    private func createImage(from data: Data, width: Int, height: Int) -> UIImage? {
        let bytesPerPixel = 4
        let bytesPerRow = width * bytesPerPixel
        
        guard let provider = CGDataProvider(data: data as CFData),
              let cgImage = CGImage(
                width: width,
                height: height,
                bitsPerComponent: 8,
                bitsPerPixel: bytesPerPixel * 8,
                bytesPerRow: bytesPerRow,
                space: CGColorSpaceCreateDeviceRGB(),
                bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedLast.rawValue),
                provider: provider,
                decode: nil,
                shouldInterpolate: false,
                intent: .defaultIntent
              ) else {
            return nil
        }
        
        return UIImage(cgImage: cgImage)
    }
    // --- Native Animation Support ---
    
    private var displayLink: CADisplayLink?
    private var animationManager: AnimationManager?
    
    func startAnimation(config: [String: Any]) {
        stopAnimation()
        
        if let type = config["type"] as? String, type == "confetti" {
            animationManager = ConfettiAnimation(config: config)
            
            displayLink = CADisplayLink(target: self, selector: #selector(animationLoop))
            displayLink?.add(to: .main, forMode: .common)
        }
    }
    
    func stopAnimation() {
        displayLink?.invalidate()
        displayLink = nil
        animationManager = nil
    }
    
    @objc private func animationLoop() {
        guard let manager = animationManager else { return }
        
        // Create image context for drawing
        UIGraphicsBeginImageContextWithOptions(bounds.size, false, 0.0)
        guard let context = UIGraphicsGetCurrentContext() else { return }
        
        // Clear
        context.clear(bounds)
        
        // Update and draw
        manager.updateAndDraw(context: context, size: bounds.size)
        
        // Get image and update view
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        imageView?.image = image
        
        // Auto-stop
        if manager.isFinished {
            stopAnimation()
        }
    }
}

// --- Animation Classes ---

protocol AnimationManager {
    func updateAndDraw(context: CGContext, size: CGSize)
    var isFinished: Bool { get }
}

class ConfettiAnimation: AnimationManager {
    private var particles: [Particle] = []
    
    // Config
    private let particleCount: Int
    private let startVelocity: CGFloat
    private let spread: CGFloat
    private let angle: CGFloat
    private let gravity: CGFloat
    private let drift: CGFloat
    private let decay: CGFloat
    private let colors: [UIColor]
    private let scalar: CGFloat
    
    private var initialized = false
    
    init(config: [String: Any]) {
        self.particleCount = config["particleCount"] as? Int ?? 50
        self.startVelocity = (config["startVelocity"] as? CGFloat) ?? 45.0
        self.spread = (config["spread"] as? CGFloat) ?? 45.0
        self.angle = (config["angle"] as? CGFloat) ?? 90.0
        self.gravity = (config["gravity"] as? CGFloat) ?? 1.0
        self.drift = (config["drift"] as? CGFloat) ?? 0.0
        self.decay = (config["decay"] as? CGFloat) ?? 0.9
        self.scalar = (config["scalar"] as? CGFloat) ?? 1.0
        
        if let colorValues = config["colors"] as? [Int] {
            self.colors = colorValues.map { UIColor(rgb: $0) }
        } else {
            self.colors = [.red, .blue]
        }
        
        // Initialize particles
        for _ in 0..<particleCount {
            particles.append(createParticle())
        }
    }
    
    func updateAndDraw(context: CGContext, size: CGSize) {
        if !initialized {
            let startX = size.width / 2
            let startY = size.height / 2
            for i in 0..<particles.count {
                particles[i].x = startX
                particles[i].y = startY
            }
            initialized = true
        }
        
        var activeParticles = 0
        
        for i in 0..<particles.count {
            // Skip dead particles
            if particles[i].dead { continue }
            
            // Physics
            particles[i].x += particles[i].vx
            particles[i].y += particles[i].vy
            particles[i].vy += gravity
            particles[i].vx *= decay
            particles[i].vy *= decay
            particles[i].x += drift
            
            // Draw
            context.setFillColor(particles[i].color.cgColor)
            context.fillEllipse(in: CGRect(
                x: particles[i].x - particles[i].radius,
                y: particles[i].y - particles[i].radius,
                width: particles[i].radius * 2,
                height: particles[i].radius * 2
            ))
            
            // Check bounds
            if particles[i].y > size.height + 50 {
                particles[i].dead = true
            } else {
                activeParticles += 1
            }
        }
        
        // Auto-stop if all particles are dead (we can't set isActive directly on struct protocol, 
        // so we'll return a status or handle it in the view. 
        // For now, let's assume the view checks a property or we add one to the protocol)
    }
    
    var isFinished: Bool {
        return initialized && particles.allSatisfy { $0.dead }
    }
    
    private func createParticle() -> Particle {
        let randomAngle = (angle - spread / 2 + CGFloat.random(in: 0...1) * spread) * (.pi / 180)
        let speed = startVelocity * (0.5 + CGFloat.random(in: 0...1) * 0.5)
        
        return Particle(
            x: 0, // Set in update
            y: 0,
            vx: cos(randomAngle) * speed,
            vy: -sin(randomAngle) * speed,
            color: colors.randomElement() ?? .red,
            radius: (3 + CGFloat.random(in: 0...1) * 4) * scalar,
            dead: false
        )
    }
}

struct Particle {
    var x: CGFloat
    var y: CGFloat
    var vx: CGFloat
    var vy: CGFloat
    var color: UIColor
    var radius: CGFloat
    var dead: Bool
}

extension UIColor {
    convenience init(rgb: Int) {
        self.init(
            red: CGFloat((rgb >> 16) & 0xFF) / 255.0,
            green: CGFloat((rgb >> 8) & 0xFF) / 255.0,
            blue: CGFloat(rgb & 0xFF) / 255.0,
            alpha: CGFloat((rgb >> 24) & 0xFF) / 255.0
        )
    }
}

class DCFCanvasComponent: NSObject, DCFComponent {
    
    // Track all canvas views by canvasId for tunnel method routing
    private static var canvasViews: [String: DCFCanvasView] = [:]
    
    override required init() { super.init() }
    
    func createView(props: [String: Any]) -> UIView {
        let view = DCFCanvasView()
        configureView(view, with: props)
        return view
    }
    
    func updateView(_ view: UIView, withProps props: [String: Any]) -> Bool {
        guard let canvasView = view as? DCFCanvasView else { return false }
        configureView(canvasView, with: props)
        return true
    }
    
    private func configureView(_ view: DCFCanvasView, with props: [String: Any]) {
        // Register canvas by ID for tunnel method routing
        if let canvasId = props["canvasId"] as? String {
            view.canvasId = canvasId
            DCFCanvasComponent.canvasViews[canvasId] = view
        }
        
        // Handle Command Pattern
        if let command = props["canvasCommand"] as? [String: Any] {
            if let name = command["name"] as? String {
                if name == "startAnimation" {
                    if let config = command["config"] as? [String: Any] {
                        view.startAnimation(config: config)
                    }
                } else if name == "stopAnimation" {
                    view.stopAnimation()
                }
            }
        }
        
        // Everything else is handled in Skia-land via texture updates
        // No hardcoded native primitives (particles, shaders, etc.)
    }
    
    func applyLayout(_ view: UIView, layout: YGNodeLayout) {
        view.frame = CGRect(
            x: layout.left,
            y: layout.top,
            width: layout.width,
            height: layout.height
        )
    }
    
    func getIntrinsicSize(_ view: UIView, forProps props: [String: Any]) -> CGSize {
        return CGSize(width: UIView.noIntrinsicMetric, height: UIView.noIntrinsicMetric)
    }
    
    func viewRegisteredWithShadowTree(_ view: UIView, nodeId: String) {
        // Set nodeId using objc_setAssociatedObject for propagateEvent to work
        objc_setAssociatedObject(view, 
                               UnsafeRawPointer(bitPattern: "viewId".hashValue)!, 
                               nodeId, 
                               .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        
        if let canvasView = view as? DCFCanvasView {
            canvasView.nodeId = nodeId
        }
    }
    
    func prepareForRecycle(_ view: UIView) {
        if let canvasView = view as? DCFCanvasView {
            canvasView.nodeId = nil
            canvasView.stopAnimation() // Ensure animation stops
            
            // Unregister from tracking
            if let canvasId = canvasView.canvasId {
                DCFCanvasComponent.canvasViews.removeValue(forKey: canvasId)
            }
        }
    }
    
    static func handleTunnelMethod(_ method: String, params: [String: Any]) -> Any? {
        // Handle updatePixels tunnel for Flutter/Skia texture rendering
        if method == "updatePixels" {
            guard let canvasId = params["canvasId"] as? String,
                  let canvasView = canvasViews[canvasId],
                  let pixels = params["pixels"] as? FlutterStandardTypedData,
                  let width = params["width"] as? Int,
                  let height = params["height"] as? Int else {
                print("❌ Canvas: Missing required params for updatePixels")
                return false
            }
            
            canvasView.updateTexture(with: pixels.data, width: width, height: height)
            return true
        }
        
        return nil
    }
}

File: packages/dcf_reanimated/android/src/main/kotlin/com/dotcorr/dcf_reanimated/components/DCFCanvasComponent.kt
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.dotcorr.dcf_reanimated.components

import android.content.Context
import android.graphics.PointF
import android.graphics.SurfaceTexture
import android.view.TextureView
import android.view.View
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Color
import com.dotcorr.dcflight.components.DCFComponent
import com.dotcorr.dcflight.components.DCFNodeLayout
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.sqrt
import kotlin.math.cos
import kotlin.math.sin

class DCFCanvasComponent : DCFComponent() {
    private var canvasView: DCFCanvasView? = null

    override fun createView(context: Context, props: Map<String, Any?>): View {
        canvasView = DCFCanvasView(context)
        
        // Set initial size from props if available
        val width = (props["width"] as? Number)?.toInt() ?: 0
        val height = (props["height"] as? Number)?.toInt() ?: 0
        if (width > 0 && height > 0) {
            val layoutParams = android.view.ViewGroup.LayoutParams(width, height)
            canvasView!!.layoutParams = layoutParams
        }
        
        updateView(canvasView!!, props)
        return canvasView!!
    }

    override fun updateView(view: View, props: Map<String, Any?>): Boolean {
        if (view is DCFCanvasView) {
            view.update(props)
            return true
        }
        return false
    }
    
    override fun getIntrinsicSize(view: View, props: Map<String, Any>): PointF {
        // Canvas size is determined by layout or props, not intrinsic content
        val width = (props["width"] as? Number)?.toFloat() ?: 0f
        val height = (props["height"] as? Number)?.toFloat() ?: 0f
        return PointF(width, height)
    }
    
    override fun applyLayout(view: View, layout: DCFNodeLayout) {
        // Apply Yoga layout to the view
        view.layout(
            layout.left.toInt(),
            layout.top.toInt(),
            (layout.left + layout.width).toInt(),
            (layout.top + layout.height).toInt()
        )
        // Notify canvas view of layout size change
        if (view is DCFCanvasView) {
            view.onLayoutApplied(layout.width, layout.height)
        }
    }
    
    override fun viewRegisteredWithShadowTree(view: View, nodeId: String) {
        // No special handling needed for canvas registration
    }

    // Handle tunnel methods from Dart
    override fun handleTunnelMethod(method: String, arguments: Map<String, Any?>): Any? {
        return when (method) {
            "updatePixels" -> {
                // Dart sends: canvasId, pixels (ByteArray), width, height
                val canvasId = arguments["canvasId"] as? String
                val pixels = arguments["pixels"] as? ByteArray
                val width = (arguments["width"] as? Number)?.toInt()
                val height = (arguments["height"] as? Number)?.toInt()

                if (canvasId != null && pixels != null && width != null && height != null) {
                    val view = DCFCanvasView.canvasViews[canvasId]
                    if (view != null) {
                        view.updatePixels(pixels, width, height)
                        true
                    } else {
                        false  // View not ready yet (retry later)
                    }
                } else {
                    null  // Invalid params
                }
            }
            else -> null  // Method not supported
        }
    }
}

class DCFCanvasView(context: Context) : TextureView(context), TextureView.SurfaceTextureListener {
    companion object {
        val canvasViews = ConcurrentHashMap<String, DCFCanvasView>()
    }

    private var canvasId: String? = null
    private var surfaceTexture: SurfaceTexture? = null

    init {
        surfaceTextureListener = this
        // Transparent background - canvas content comes from texture
        isOpaque = false
        // Allow touches to pass through to views behind the canvas
        isClickable = false
        isFocusable = false
    }

    fun update(props: Map<String, Any?>) {
        val id = props["canvasId"] as? String
        if (id != null && canvasId != id) {
            // Unregister old ID
            if (canvasId != null) {
                canvasViews.remove(canvasId!!)
            }
            canvasId = id
            canvasViews[id!!] = this
            android.util.Log.d("DCFCanvasView", "Registered canvasId: $id")
        }
        
        // Handle Command Pattern
        val command = props["canvasCommand"] as? Map<String, Any>
        if (command != null) {
            handleCommand(command)
        }
        
        // Note: Size is now handled by Yoga layout via applyLayout
        // We don't set layoutParams here anymore - Yoga handles it
    }
    
    fun onLayoutApplied(width: Float, height: Float) {
        android.util.Log.d("DCFCanvasView", "onLayoutApplied width: $width height: $height, canvasId: $canvasId")
        // Ensure we're registered if we have a valid size
        if (width > 0 && height > 0 && canvasId != null) {
            canvasViews[canvasId!!] = this
        }
    }
    fun updatePixels(pixels: ByteArray, width: Int, height: Int) {
        val surface = surfaceTexture ?: return
        // In a real implementation, we would update the SurfaceTexture here.
        // Since SurfaceTexture requires an OpenGL texture ID, we would typically
        // use EGL to bind the texture and update it with the pixels.
        // For this implementation, we will use a simpler approach:
        // Draw the pixels to a Bitmap and then to the Surface.
        
        try {
            // Convert RGBA to ARGB (Android Bitmap expects ARGB_8888)
            // Use IntArray for setPixels which is more reliable than ByteBuffer
            val argbPixels = IntArray(width * height)
            for (y in 0 until height) {
                for (x in 0 until width) {
                    val srcIndex = (y * width + x) * 4
                    if (srcIndex + 3 < pixels.size) {
                        // RGBA: R, G, B, A (from Dart)
                        // ARGB: A, R, G, B (for Android Bitmap)
                        val r = pixels[srcIndex].toInt() and 0xFF
                        val g = pixels[srcIndex + 1].toInt() and 0xFF
                        val b = pixels[srcIndex + 2].toInt() and 0xFF
                        val a = pixels[srcIndex + 3].toInt() and 0xFF
                        
                        // Pack into ARGB int: (A << 24) | (R << 16) | (G << 8) | B
                        argbPixels[y * width + x] = (a shl 24) or (r shl 16) or (g shl 8) or b
                    }
                }
            }
            
            val bitmap = android.graphics.Bitmap.createBitmap(width, height, android.graphics.Bitmap.Config.ARGB_8888)
            bitmap.setPixels(argbPixels, 0, width, 0, 0, width, height)
            
            val canvas = lockCanvas()
            if (canvas != null) {
                // Clear canvas first to avoid artifacts
                canvas.drawColor(android.graphics.Color.TRANSPARENT, android.graphics.PorterDuff.Mode.CLEAR)
                canvas.drawBitmap(bitmap, 0f, 0f, null)
                unlockCanvasAndPost(canvas)
            }
        } catch (e: Exception) {
            android.util.Log.e("DCFCanvasView", "Error updating texture", e)
        }
    }
    
    // NEW: Command-based rendering (Phase 3) - executes Skia commands directly
    fun updateCommands(commands: List<Map<String, Any>>, width: Int, height: Int): Boolean {
        if (!isAvailable) return false

        val canvas = lockCanvas() ?: return false
        
        try {
            canvas.drawColor(android.graphics.Color.TRANSPARENT, android.graphics.PorterDuff.Mode.CLEAR)
            executeCommands(commands, canvas, width, height)
            unlockCanvasAndPost(canvas)
            return true
        } catch (e: Exception) {
            android.util.Log.e("DCFCanvasView", "Error executing commands", e)
            unlockCanvasAndPost(canvas)
            return false
        }
    }
    
    // Execute drawing commands on Android Canvas (Skia)
    private fun executeCommands(commands: List<Map<String, Any>>, canvas: Canvas, width: Int, height: Int) {
        val paint = Paint().apply { isAntiAlias = true }
        
        for (command in commands) {
            val type = command["type"] as? String ?: continue
            @Suppress("UNCHECKED_CAST")
            val params = command["params"] as? Map<String, Any> ?: continue
            
            when (type) {
                "drawRect" -> {
                    val rect = params["rect"] as? List<Double>
                    val paintData = params["paint"] as? Map<String, Any>
                    if (rect != null && rect.size == 4 && paintData != null) {
                        applyPaint(paint, paintData)
                        canvas.drawRect(
                            rect[0].toFloat(), rect[1].toFloat(),
                            rect[2].toFloat(), rect[3].toFloat(),
                            paint
                        )
                    }
                }
                // ... (other commands omitted for brevity, logic remains same)
            }
        }
    }
    
    private fun applyPaint(paint: Paint, paintData: Map<String, Any>) {
        val color = (paintData["color"] as? Number)?.toInt()
        if (color != null) paint.color = color
        
        val style = (paintData["style"] as? Number)?.toInt()
        if (style != null) {
            paint.style = when (style) {
                0 -> Paint.Style.FILL
                1 -> Paint.Style.STROKE
                else -> Paint.Style.FILL
            }
        }
        
        val strokeWidth = (paintData["strokeWidth"] as? Number)?.toFloat()
        if (strokeWidth != null) paint.strokeWidth = strokeWidth
        
        val isAntiAlias = paintData["isAntiAlias"] as? Boolean
        if (isAntiAlias != null) paint.isAntiAlias = isAntiAlias
    }

    // --- Native Animation Support ---

    private var animationManager: AnimationManager? = null
    private val frameCallback = object : android.view.Choreographer.FrameCallback {
        override fun doFrame(frameTimeNanos: Long) {
            if (animationManager?.isActive == true) {
                val canvas = lockCanvas()
                if (canvas != null) {
                    try {
                        canvas.drawColor(android.graphics.Color.TRANSPARENT, android.graphics.PorterDuff.Mode.CLEAR)
                        animationManager?.updateAndDraw(canvas, width, height)
                    } finally {
                        unlockCanvasAndPost(canvas)
                    }
                }
                android.view.Choreographer.getInstance().postFrameCallback(this)
            }
        }
    }

    fun startAnimation(config: Map<String, Any>) {
        stopAnimation() // Stop existing
        
        val type = config["type"] as? String
        if (type == "confetti") {
            animationManager = ConfettiAnimation(config)
            android.view.Choreographer.getInstance().postFrameCallback(frameCallback)
        }
    }

    // Old startAnimation and stopAnimation for animationManager are removed.
    // The new animation system uses ConfettiView directly.

    private var confettiView: ConfettiView? = null

    fun handleCommand(command: Map<String, Any>) {
        val type = command["type"] as? String ?: return
        
        when (type) {
            "confetti" -> {
                stopAnimation()
                // Create new HW accelerated view
                // Command map IS the config now
                confettiView = ConfettiView(context, command)
                addView(confettiView, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
            }
            "clear" -> stopAnimation()
            // Future: "drawShape" -> ...
        }
    }

    fun stopAnimation() {
        if (confettiView != null) {
            removeView(confettiView)
            confettiView?.stop()
            confettiView = null
        }
    }

    // The original SurfaceTextureListener methods are removed.
    // override fun onSurfaceTextureAvailable(surface: SurfaceTexture, width: Int, height: Int) {
    //     this.surfaceTexture = surface
    // }
    //
    // override fun onSurfaceTextureSizeChanged(surface: SurfaceTexture, width: Int, height: Int) {}
    //
    // override fun onSurfaceTextureDestroyed(surface: SurfaceTexture): Boolean {
    //     stopAnimation()
    //     return true
    // }
    //
    // override fun onSurfaceTextureUpdated(surface: SurfaceTexture) {}
}

// --- Animation Classes ---
// The original AnimationManager and ConfettiAnimation classes are removed.

class ConfettiView(context: Context, config: Map<String, Any>) : View(context) {
    private val particles = ArrayList<Particle>()
    private val paint = Paint()
    private var isActive = true
    private val gravity = 0.5f
    private val decay = 0.95f
    private val drift = 0.0f
    
    // Config
    private val scalar = (config["scalar"] as? Number)?.toFloat() ?: 1.0f
    private val spread = (config["spread"] as? Number)?.toFloat() ?: 60.0f
    private val startVelocity = (config["startVelocity"] as? Number)?.toFloat() ?: 45.0f
    private val colors = (config["colors"] as? List<String>)?.map { Color.parseColor(it) } ?: listOf(Color.RED, Color.GREEN, Color.BLUE)
    private val elementCount = (config["elementCount"] as? Number)?.toInt() ?: 50

    init {
        // Initialize particles
        // We need width/height which we don't have yet in init
        // So we'll spawn them in onDraw or onSizeChanged
    }
    
    private var initialized = false

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        
        if (!isActive) return

        if (!initialized) {
            val startX = width / 2f
            val startY = height / 2f
            
            for (i in 0 until elementCount) {
                particles.add(createParticle(startX, startY))
            }
            initialized = true
        }

        var activeParticles = 0
        
        // Loop through particles
        // Using indices to avoid iterator allocation in draw loop
        for (i in particles.indices) {
            val p = particles[i]
            if (p.dead) continue

            // Physics
            p.x += p.vx
            p.y += p.vy
            p.vy += gravity
            p.vx *= decay
            p.vy *= decay
            p.x += drift

            // Draw
            paint.color = p.color
            canvas.drawCircle(p.x, p.y, p.radius, paint)

            // Check bounds
            if (p.y > height + 50) {
                p.dead = true
            } else {
                activeParticles++
            }
        }

        if (activeParticles > 0) {
            invalidate() // Schedule next frame (HW accelerated)
        } else if (initialized) {
            isActive = false // Auto-stop
        }
    }
    
    fun stop() {
        isActive = false
    }

    private fun createParticle(startX: Float, startY: Float): Particle {
        val randomAngle = (Math.random() * spread - spread / 2 - 90) * (Math.PI / 180) // -90 for Up
        val speed = startVelocity * (0.5 + Math.random() * 0.5).toFloat()
        
        val p = Particle()
        p.x = startX
        p.y = startY
        p.vx = (Math.cos(randomAngle) * speed).toFloat()
        p.vy = (Math.sin(randomAngle) * speed).toFloat()
        p.color = colors[(Math.random() * colors.size).toInt()]
        p.radius = ((3 + Math.random() * 4) * scalar).toFloat()
        return p
    }
}

class Particle {
    var x: Float = 0f
    var y: Float = 0f
    var vx: Float = 0f
    var vy: Float = 0f
    var color: Int = 0
    var radius: Float = 0f
    var dead: Boolean = false
}

